not-script/not-script.c

```diff
@@ -0,0 +1,533 @@
+#define BACKEND_VBD "backend/vbd/"
+#define ARGV2_PREFIX (XENBUS_PATH_PREFIX BACKEND_VBD)
+
+    const char *xs_path_raw = getenv("XENBUS_PATH");
+
+    if (argc < 2 || argc > 3)
+        errx(1, "Usage: [add|remove] [XENBUS_PATH=backend/vbd/REMOTE_DOMAIN/ID] (got %d arguments, expected 2 or 3)", argc);
+
+    if (argc >= 3) {
+        if (strncmp(argv[2], ARGV2_PREFIX, sizeof(ARGV2_PREFIX) - 1))
+            errx(1, "Second argument must begin with XENBUS_PATH=backend/vbd/");
+        const char *const new_path = argv[2] + (sizeof(XENBUS_PATH_PREFIX) - 1);
+
+        if ((xs_path_raw != NULL) && (strcmp(xs_path_raw, new_path) != 0))
+            errx(1, "XENBUS_PATH was passed both on the command line and in"
+                    " the environment, but the values were different");
+
+        xs_path_raw = new_path;
+    } else if (xs_path_raw == NULL) {
+        errx(1, "Xenstore path required");
+    } else if (strncmp(xs_path_raw, BACKEND_VBD, sizeof(BACKEND_VBD) - 1)) {
+        errx(1, "Bad Xenstore path %s", xs_path_raw);
+    }
+
+    if (strcmp(argv[1], "add") == 0) {
+        action = ADD;
+    } else if (strcmp(argv[1], "remove") == 0) {
+        action = REMOVE;
+    } else {
+        errx(1, "Bad command (expected \"add\" or \"remove\")");
+    }
+
+    const char *const xs_path = xs_path_raw;
+    size_t xs_path_len;
+
+    {
+        /* strtoul() is not const-correct, sorry... */
+        char *xs_path_extra = (char *)(xs_path + (sizeof(BACKEND_VBD) - 1));
+        unsigned long peer_domid, tmp;
+        validate_int_start(&xs_path_extra, &peer_domid, true);
+        if (*xs_path_extra++ != '/')
+            errx(1, "Peer domain ID %lu not followed by '/'", peer_domid);
+        validate_int_start(&xs_path_extra, &tmp, true);
+        if (*xs_path_extra)
+            errx(1, "Junk after XenStore device ID %lu", tmp);
+        if (peer_domid >= DOMID_FIRST_RESERVED)
+            errx(1, "Peer domain ID %lu too large (limit %d)", peer_domid, DOMID_FIRST_RESERVED);
+        xs_path_len = (size_t)(xs_path_extra - xs_path);
+    }
+
+    struct xs_handle *const h = xs_open(0);
+    if (!h)
+        err(1, "Cannot connect to XenStore");
+
+    size_t const buffer_size = xs_path_len + sizeof("/physical-device-path");
+    char *const xenstore_path_buffer = malloc(buffer_size);
+    if (!xenstore_path_buffer)
+        err(1, "malloc()");
+    memcpy(xenstore_path_buffer, xs_path, xs_path_len);
+    xenstore_path_buffer[xs_path_len] = '/';
+    /* Buffer to copy extra data into */
+    char *const extra_path = xenstore_path_buffer + xs_path_len + 1;
+    unsigned int len, path_len;
+    bool const autoclear = get_opened(h, extra_path, xenstore_path_buffer,
+                                      action == REMOVE ? '1' : '0');
+
+    if (action == REMOVE) {
+        if (!autoclear)
+            remove_device(h, xenstore_path_buffer, extra_path);
+        return 0;
+    }
+
+    strcpy(extra_path, "params");
```

Comment:
Add static assert (here) that it will fit? Currently it relies on `physical-device-path` being the longest path, but there is no compile-time check for that - especially important to have a pattern to follow when adding more xenstore writes.

This applies to all the other writes to `extra_path` too.